URL.parse() static method docs (#33207)
mdn/content: files/en-us/web/api/url/parse_static/index.md

---
title: "URL: parse() static method"
short-title: parse()
slug: Web/API/URL/parse_static
page-type: web-api-static-method
browser-compat: api.URL.parse_static
---

{{ApiRef("URL API")}}

The **`URL.parse()`** static method of the {{domxref("URL")}} interface returns a newly created {{domxref("URL")}} object representing the URL defined by the parameters.

If the given base URL or the resulting URL are not parsable and valid URLs, `null` is returned.
This is an alternative to using the normal constructor to construct a `URL` within a [try...catch](/en-US/docs/Web/JavaScript/Reference/Statements/try...catch) block, or using {{domxref("URL.canParse_static", "canParse()")}} to check the parameters and returning `null` if the method returns `false`.

## Syntax

```js-nolint
URL.parse(url)
URL.parse(url, base)
```

### Parameters

- `url`

>   - : A string or any other object with a {{Glossary("stringifier")}} — including, for example, an {{htmlelement("a")}} or {{htmlelement("area")}} element — that represents an absolute or relative URL.

Author: No, I would prefer we were accurate.

```suggestion
  - : A string or any other object with a {{Glossary("stringifier")}} — including, for example, an {{domxref("HTMLAnchorElement")}} or {{domxref("HTMLAreaElement")}} element — that represents an absolute or relative URL.
```

    If `url` is a relative URL, `base` is required, and will be used as the base URL.
    If `url` is an absolute URL, a given `base` will be ignored.
- `base` {{optional_inline}}
  - : A string representing the base URL to use in cases where `url` is a relative URL.
    If not specified, it defaults to `undefined`.

> **Note:** The `url` and `base` arguments will each be stringified from whatever value you pass, just like with other Web APIs that accept a string.

> > In particular, you can use an existing {{domxref("URL")}} object for either argument, and it will be stringified to the object's {{domxref("URL.href", "href")}} property.

Author: We are a bit. I've modified the `url` and this so that the examples are in the  note. That way the `url` section is more concise and this note reads as useful additional information.

### Return value

A `URL` if the parameters can be resolved to a valid URL; `null` otherwise.

## Examples

This live example demonstrates how to use the `URL.parse()` static method for a few different absolute and relative URL values.

```html hidden
<pre id="log"></pre>
```

```css hidden
#log {
  height: 250px;
  overflow: scroll;
  padding: 0.5rem;
  border: 1px solid black;
}
```

```js hidden
const logElement = document.getElementById("log");
function log(text) {
  logElement.innerText += `${text}\n`;
}
```

First we check that the `URL.parse()` method is supported using the condition `"parse" in URL`.
If the method is supported we log the result of checking a relative URL with a base URL, a valid absolute URL with a valid base URL (which gets ignored), and an invalid base URL that returns `null`.

We also log the case when `URL.parse()` is not supported.

```js
if ("parse" in URL) {
  let url = "en-US/docs";
  let base = "https://developer.mozilla.org";
  let result = URL.parse(url, base);

>   log("//Relative URL appended to valid base URL");

Author: @wbamberg
1. I have removed comments in the logs. 
   - The purpose of that was to provide clear relationship between the javascript and the log output - because by the time you got to the output it isn't so clear what the log is showing. 
   - This is now fixed by using literals (as you suggested) which reduces the separation, and using numbered log items to match the numbered code.
2. Yes, this makes sense. I don't really don't want to do it because of the additional thinking/effort :-(
3. Fixed up all the `Url .`=> `URL`. Sorry.
4. 
Appreciate all the advice.

>   log(`Url.parse("${url}", "${base}")\n// ${result}`);

Author: Thanks. Good point.

  url = "https://example.org/some/docs";
  result = URL.parse(url, base);
  log("\n//Absolute URL (base URL ignored)");
  log(`Url.parse("${url}", "${base}")\n// => ${result}`);

  log("\n//Invalid base URL (missing colon)");
  base = "https//developer.mozilla.org";
  result = URL.parse(url, base);
  log(`Url.Parse("${url}", "${base}")\n// ${result}`);
} else {
  log("URL.parse() not supported");
}
```

Last of all, the code below shows that the `baseUrl` doesn't have to be a string.
Here we have passed a `URL` object.

```js
if ("parse" in URL) {

>   log("\nRelative URL with base URL supplied as a URL object");

Reviewer: ```suggestion
  log("\n Relative URL with base URL supplied as a URL object");
```

  base = new URL("https://developer.mozilla.org/");
  url = "/en-US/docs";
  result = URL.parse(url, base);
  log(`Url.parse("${url}", "${base}")\n// ${result}`);
}
```

The results of each of the checks are shown below.

{{EmbedLiveSample('Examples', '100%', '300')}}

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

## See also

- [`URL()` constructor](/en-US/docs/Web/API/URL/URL), which throws if the passed parameters define an invalid URL

> - [A polyfill of `URL.canParse()`](https://github.com/zloirock/core-js#url-and-urlsearchparams) is available in [`core-js`](https://github.com/zloirock/core-js)

Reviewer: Why link to a polyfill for a different method?

Author: Copy paste error

> - [`URL()` constructor](/en-US/docs/Web/API/URL/URL), which throws if the passed parameters define an invalid URL
> - [A polyfill of `URL.canParse()`](https://github.com/zloirock/core-js#url-and-urlsearchparams) is available in [`core-js`](https://github.com/zloirock/core-js)

Author: ```suggestion
- [`URL()` constructor](/en-US/docs/Web/API/URL/URL), which throws if the passed parameters define an invalid URL
```

Reviewer 2: `core-js` contains `URL.parse` polyfill by this link, just the title should be renamed.

Reviewer 2: @hamishwillee

```js
if ("parse" in URL) {
  let result = URL.parse("en-US/docs", "https://developer.mozilla.org");
  log("//Relative URL appended to valid base URL");
  log(`Url.parse("${url}", "${base}")\n// ${result}`);

  url = "https://example.org/some/docs";
  result = URL.parse(url, base);
  log("\n//Absolute URL (base URL ignored)");
  log(`Url.parse("${url}", "${base}")\n// => ${result}`);

  log("\n//Invalid base URL (missing colon)");
  base = "https//developer.mozilla.org";
  result = URL.parse(url, base);
  log(`Url.Parse("${url}", "${base}")\n// ${result}`);
} else {
  log("URL.parse() not supported");
}
```

Last of all, the code below shows that the `baseUrl` doesn't have to be a string.
Here we have passed a `URL` object.

```js
if ("parse" in URL) {
  log("\nRelative URL with base URL supplied as a URL object");
  base = new URL("https://developer.mozilla.org/");
  url = "/en-US/docs";
  result = URL.parse(url, base);
  log(`Url.parse("${url}", "${base}")\n// ${result}`);
}
```

The results of each of the checks are shown below.

{{EmbedLiveSample('URL.parse()', '100%', '300')}}

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

## See also

> - [`URL()` constructor](/en-US/docs/Web/API/URL/URL), which throws if the passed parameters define an invalid URL

Reviewer 2: https://github.com/mdn/content/pull/33207#discussion_r1582459147
```suggestion
- [`URL()` constructor](/en-US/docs/Web/API/URL/URL), which throws if the passed parameters define an invalid URL
- [A polyfill of `URL.parse()`](https://github.com/zloirock/core-js#url-and-urlsearchparams) is available in [`core-js`](https://github.com/zloirock/core-js)
```

Author: Thanks!

---

{{ApiRef("URL API")}}

The **`URL.parse()`** static method of the {{domxref("URL")}} interface returns a newly created {{domxref("URL")}} object representing the URL defined by the parameters.

If the given base URL or the resulting URL are not parsable and valid URLs, `null` is returned.
This is an alternative to using the {{domxref("URL.URL", "URL()")}} constructor to construct a `URL` within a [try...catch](/en-US/docs/Web/JavaScript/Reference/Statements/try...catch) block, or using {{domxref("URL.canParse_static", "canParse()")}} to check the parameters and returning `null` if the method returns `false`.

## Syntax

```js-nolint
URL.parse(url)
URL.parse(url, base)
```

### Parameters

- `url`
  - : A string or any other object with a {{Glossary("stringifier")}} that represents an absolute or relative URL.
    If `url` is a relative URL, `base` is required, and is used to resolve the final absolute URL.
    If `url` is an absolute URL, a given `base` will not be used to create the resulting URL.
- `base` {{optional_inline}}

  - : A string representing the base URL to use in cases where `url` is a relative URL.
    If not specified, it defaults to `undefined`.

>     When specify a `base` URL, the resolved URL is not simply a concatenation of `url` and `base`.

Reviewer: ```suggestion
    When you specify a `base` URL, the resolved URL is not simply a concatenation of `url` and `base`.
```

>     When specify a `base` URL, the resolved URL is not simply a concatenation of `url` and `base`.

Reviewer: ```suggestion
    When you specify a `base` URL, the resolved URL is not simply a concatenation of `url` and `base`.
```

>     Parent-relative and current-directory-relative URLS are relative to the current directory of the `base` URL, which includes only path segments up until the last forward-slash, but not any after.

Reviewer: ```suggestion
    Parent-relative and current-directory-relative URLs are relative to the current directory of the `base` URL, which includes only path segments up until the last forward-slash, but not any after.
```
